### ISSUE TITLE
Check keywords in curry __call__ to properly re-raise TypeErrors

### ISSUE THREAD

**Issue by jrderuiter** [pytoolz/toolz#288]
I am sometimes running into issues where TypeErrors are not properly re-raised from curry's __call__ method, resulting in returning the curried function (as if the function does not have sufficient arguments) instead of actually executing the function. 

I think this is because I supply extra arguments as keyword arguments, rather than positional arguments. As far as I can see, the code in __call__ that is called when a TypeError is raised by the partial, only checks positional arguments to determine if the curried function has sufficient arguments and should therefore re-raise the TypeError. In my case, because most required arguments are passed as keyword arguments, this test counts too few arguments and therefore does not raise a TypeError. 

Shouldn't this behaviour be changed to take the keyword arguments into account?

This is the code I'm referring to:
```{python}
    def __call__(self, *args, **kwargs):
        try:
            return self._partial(*args, **kwargs)
        except TypeError:
            # If there was a genuine TypeError
            required_args = _num_required_args(self.func)
            if (required_args is not None and
                    len(args) + len(self.args) >= required_args):
                raise

        return curry(self._partial, *args, **kwargs)
```

Shouldn't it be more like this?

```{python}
    def __call__(self, *args, **kwargs):
        try:
            return self._partial(*args, **kwargs)
        except TypeError:
            # If there was a genuine TypeError
            required_args = _num_required_args(self.func)
            if (required_args is not None and
                    len(args) + len(self.args) + len(self.keywords) >= required_args):
                raise

        return curry(self._partial, *args, **kwargs)
```

This naive change gives the required results in my simple case, though I don't know if I'm missing other behaviour here.

**Comment by mrocklin** [comment]
It's tricky because not all keywords necessarily count towards the number of required arguments.

Sorry for the long delay in response.  I'm a bit busy this month.

**Comment by jrderuiter** [comment]
I have an updated piece of code that should take that into account. It
effectively determines the required arguments and checks for overlap in
given keywords and the required arguments. The remaining 'unfulfilled'
arguments are then expected to come from args. Will put the code up later
this week.

**Comment by eriknw** [comment]
This is indeed potentially tricky and messy to support Python 2.6 through 3.5, but it's about time we properly address this for normal Python functions, so thanks for your help!

I have an idea.  Instead of performing introspection to try to determine whether a function call is valid or not, we can create a new function (that does nothing) with the same signature.  If it raises `TypeError` upon being called, then the call to the original function is also invalid.
```python
def mimic(func):
    """Return a new function with the same call signature as func.

    This only works on normal Python functions and not builtin functions.
    """
    noop = lambda: None
    functype = type(noop)
    try:
        c = noop.__code__
        f = func.__code__
        defaults = func.__defaults__
    except AttributeError:
        # For Python 2.6 support
        c = noop.func_code
        f = func.func_code
        defaults = func.func_defaults
    codetype = type(c)
    # Replace: argcount, nlocals, stacksize, flags, varnames, name
    args = [f.co_argcount, f.co_nlocals, f.co_stacksize, f.co_flags,
            c.co_code, c.co_consts, c.co_names, f.co_varnames,
            c.co_filename, f.co_name, c.co_firstlineno, c.co_lnotab,
            c.co_freevars, c.co_cellvars]
    if hasattr(f, 'co_kwonlyargcount'):
        # For Python 3 support
        args.insert(1, f.co_kwonlyargcount)
    newcode = codetype(*args)
    newfunc = functype(newcode, {}, f.co_name, defaults)
    newfunc.__kwdefaults__ = getattr(func, '__kwdefaults__', None)
    return newfunc
```
Although this is unfamiliar territory for most Python developers, the positive is that we defer to Python's internal system to determine the validity of a call signature.

**Comment by jrderuiter** [comment]
Interesting approach. I tried using the following myself, which essentially determines which arguments are required and checks if any keyword arguments overlap with these required arguments. Combined with *args, we then check to see if the total number of args and required keyword arguments is sufficient. In my hands this seems to work fine, though there may be corner cases I haven't thought of.

```{python}
    def __call__(self, *args, **kwargs):
        try:
            return self._partial(*args, **kwargs)
        except TypeError:
            # If there was a genuine TypeError

            # Determine number of required arguments.
            sig = inspect.signature(self.func)
            no_default = [name for name, value in sig.parameters.items()
                          if value.default == inspect.Parameter.empty]

            # Check for required arguments in keywords.
            kws = dict(kwargs)
            kws.update(self.keywords)

            req_kws = set(kws.keys()).intersection(no_default)

            if len(args) + len(self.args) + len(req_kws) >= len(no_default):
                raise

            # Handle None case?

        return curry(self._partial, *args, **kwargs)
```

Which approach would be preferred by the devs?

**Comment by jrderuiter** [comment]
I also ran into issues where I gave a wrong keyword argument, which is also currently not properly handled by curry, as it may wrongfully catch the TypeError. This suggests that the 'mimic' approach may be more suitable to handle a broader range of errors?

**Comment by eriknw** [comment]
Thanks for raising this issue and exploring solutions, @jrderuiter.  PR #297 should solve all the issues raised here.

I decided against using `mimic` so that we're not tied to the CPython interpreter.  Python 3 was able to be handled easily using `inspect.Signature.bind` and `inspect.Signature.bind_partial`, and Python 2 wasn't too painful to support.  I did end up using mimic-like functions to handle bulitins.  This discussion was very useful, and I'm glad that `curry` now has much more robust introspection!
